src/kcas.mli

```diff
@@ -87,15 +112,19 @@ module Tx : sig
       Transactions can be composed both sequentially (see {!let*}) and
       conditionally (see {!(<|>)} and {!forget}).
 
-      Transactions are performed in two phases:
+      Transactions are performed in two or three phases:
 
       1. The first phase essentially records a log of operations based on the
       {!get} and {!set} accesses of shared memory locations.
 
       2. The second phase attempts to perform the operations atomically.
 
-      Either phase may fail.  In particular, the first phase is allowed to raise
-      exceptions to signal failure.
+      3. In {!Mode.obstruction_free} a third phase verifies all read-only
+      operations.
+
+      Each phase may fail.  In particular, the first phase is allowed to raise
+      exceptions to signal failure.  Failure on the third phase raises
+      {!Mode.Interference}.
```

Comment:
This adds complexity to the otherwise pretty simple `atomically`. Is there a good reason to not just return false on interference?

Reply:
Yes, I believe there is.

See the comment in `determine_for_owner`.  It is useful to know whether failure was detected during the post hoc verification step.  It allows the caller to make a better decision on whether it makes sense to continue trying in obstruction-free mode or to switch to lock-free mode.  If failure happens before post hoc verification, switching to lock-free mode would likely not (have) help(ed).

My assumption is that `atomically` will only be used rarely by experts.  I intentionally chose the default for `atomically` and `attempt` to be `lock_free`.  The `Interference` exception will not be raised unless `obstruction_free` mode is explicitly requested.

On the other hand, `commit` takes care of handling the `Interference` exception internally.  I would assume most users will be using the `Tx` interface with `commit` and would also recommend that as `commit` also implements the retry loop and backoff (doing those correctly can be tricky enough).  Perhaps this should be made more explicit in the documentation (i.e. that `attempt` and the `Op` interface are not intended for casual users).

Also, considering the potential support for blocking transactions, there one would write the (blocking) equivalent of `commit` outside of kcas and then it would be ideal for such an implementation to be able to make the best decisions on when to switch to lock-free mode.  So, having `attempt` also raise `Interference` in case the post hoc verification failed.

Reply:
The corollary here would be that some locations are best compared in LF mode, while others in OF and a potentially more useful API would let user specify that, rather than force all compares into one mode of execution. This naturally doesn't have to come in this PR.

Reply:
> is there a good use-case for retrying in lock-free mode after failure in obstruction-mode?

I guess what I have in mind with the approach of starting, by default, in OF mode and then switching to LF mode is that it could ideally just work/perform well enough that most users would not need to think about it.

Why should it work well?  Well, if you have a relatively low contention scenario, then OF mode is likely to succeed.  So, it should make sense to go with OF mode as it would (reduce the/maintain the low) amount of writes and contention.

In a high contention scenario, OTOH, it might make sense to start in LF mode, because OF would be relatively more likely to fail and staying in OF might even cause starvation.  However, switching quickly to LF mode on interference should also mean that not much is lost by trying in OF mode first even with higher contention.  That is at least how I would intuitively assume.  It could be interesting to try to model this mathematically or through simulations.

Note that one can already explicitly start `commit` in lock-free mode and it is also possible to implement `commit` on top of `attempt` and choose desired strategy for dealing with failure and interference.

> some locations are best compared in LF mode

Hmm... So, you mean that the mode to use would be specified when creating the _location_ (rather then when calling `commit`)?  That is an interesting idea, indeed!

Hmm... Perhaps something could even be done automatically.  Perhaps something like branch-prediction.  For example, there could be some sort of memory maintained on failed compare verifies with each location.  To keep things low-overhead, that memory would only be updated in case of verify failure, but it could be consulted e.g. when creating the `CASN` descriptors (in the internal `update_as`, for example).  If the location seems "hot" (lots of compare verify failures), then the algorithm would just create a `CAS` against that location.

I think there could be interesting work here to do once we have more examples and benchmarks using this.